pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,211 @@
+// For each stream, we start a goroutine to receive the DNS policies ACKs
+// The flow of the method is as follows:
+// 1. Add the stream to the map( called by the client i.e SDP)
+// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
+// 4. Wait for the context to be done
+func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
+	stream.Send(&pb.PolicyState{})
+	return nil
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg messagehandler.DNSRequestHandler, port int, logger *slog.Logger) *FQDNDataServer {
+
+	s := &FQDNDataServer{
+		port:                port,
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+		log:                 logger.With(logfields.LogSubsys, "fqdn/server"),
+	}
+
+	return s
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	if cidr.ClusterID() != 0 {
+		s.log.Debug("Ignoring IP identity cache change for non-local CIDR", logfields.CIDR, cidr.String())
+	}
+
+	ipNet := cidr.AsIPNet()
+	if !cidr.IsSingleIP() {
+		s.log.Debug("Ignoring IP identity cache change for non-single IP CIDR", logfields.CIDR, cidr.String())
+	}
+
+	if cidr.ClusterID() == 0 && cidr.IsSingleIP() {
+		switch modType {
+		case ipcache.Upsert:
+			if oldID != nil {
+				// Remove from the old identity
+				s.deleteFromIdentityToIPLocked(oldID, ipNet.IP)
+			}
+			s.currentIdentityToIp[newID.ID] = append(s.currentIdentityToIp[newID.ID], ipNet.IP)
+
+		case ipcache.Delete:
+			if oldID != nil {
+				s.deleteFromIdentityToIPLocked(oldID, ipNet.IP)
+			}
+		}
+	}
+}
+
+// deleteFromIdentityToIPLocked deletes the given IP from the identity to IP mapping
+// It is called when the IP identity cache changes and the IP is deleted from the mapping
+// It is also called when the IP is upserted with a new identity
+// It is called with the identityToIpMutex lock held
+func (s *FQDNDataServer) deleteFromIdentityToIPLocked(identity *ipcache.Identity, ip net.IP) error {
+	if identity == nil {
+		return fmt.Errorf("identity is nil")
+	}
+
+	if ips, ok := s.currentIdentityToIp[identity.ID]; ok {
+		for i, currIp := range ips {
+			if currIp.Equal(ip) {
+				s.currentIdentityToIp[identity.ID] = slices.Delete(ips, i, i+1)
+				break
+			}
+		}
```

Comment:
Why are we not using `slices.DeleteFunc` here?

Reply:
No reason not to use it. Updated it.